### PR TITLE
LGTM 1.18: finalize change notes for C/C++

### PR DESCRIPTION
**Pull request by felicity-semmle**
As far as I know, all changes to the C/C++ queries that have customer-facing implications for the 1.18 release are merged. I've tidied up the analysis change notes so that they're ready for publication.

You probably want to review this on a commit basis, since I sorted the query table. For some reason the changes are much easier to see in the rich diff:

* [6aa6b64](https://github.com/Semmle/ql/commit/6aa6b64b813dbcf44bae11f5ea9c2cc04fdc83f50) - remove placeholder text and sort the table rows.
* [e7116f5](https://github.com/Semmle/ql/commit/e7116f57a035ac25f17a79016c374bcb359e2952) - add query identifiers for all queries.
* [2e0818d](https://github.com/Semmle/ql/commit/2e0818d8e66514f361452514736a0b997cfa4d73) - update the text to bring the notes into line with other files. Please check that I've not introduced any errors.
* [3eab1de](https://github.com/Semmle/ql/commit/3eab1de2f5af5a7413f753480af28b89c8556f15) - removed two queries that aren't available on LGTM. I'll add these to the `studio-cpp.md` file instead (c.f. [C/C++ analysis 1.17](https://help.semmle.com/wiki/pages/viewpage.action?pageId=24084582#C/C++analysis1.17-AdditionalchangesforanalysisinQLtoolsandapplicationsonly))

@jbj - do you have time to review this? Please can you ping anyone else who needs to review it.

### CODE REVIEW

File: change-notes/1.18/analysis-cpp.md [hunk 1]
```diff
@@ -1,40 +1,28 @@
...
 
-* Fixes for aggregate initializers using designators:
-  * `ClassAggregateLiteral.getFieldExpr()` previously assumed initializer expressions appeared in the same order as the declaration order of the fields, causing it to associate the expressions with the wrong fields when using designated initializers. This has been fixed.
-  * `ArrayAggregateLiteral.getElementExpr()` previously assumed initializer expressions appeared in the same order as the corresponding array elements, causing it to associate the expressions with the wrong array elements when using designated initializers. This has been fixed.
-* `Element.getEnclosingElement()` no longer includes macro accesses in its results. To explore parents and children of macro accesses, use the relevant member predicates on `MacroAccess` or `MacroInvocation`.
+* The `ClassAggregateLiteral.getFieldExpr()` and `ArrayAggregateLiteral.getElementExpr()` predicates incorrectly assumed that initializer expressions appeared in the same order as the declaration order of the fields. This resulted in the association of the expressions with the wrong fields when designated initializers were used. This has been fixed.
```
Comment (jbj) [inline change-notes/1.18/analysis-cpp.md]: There are no _fields_ in an array, so I think the merged language is confusing. Maybe replace with "elements of the aggregate" or just "elements".

File: change-notes/1.18/analysis-cpp.md [hunk 2]
```diff
@@ -1,40 +1,28 @@
...
+| Too many arguments to formatting function (`cpp/too-many-format-arguments`) | More precise results | This new query was created by splitting the old "Wrong number of arguments to formatting function" query (see row above). It reports function calls with too many arguments.  |
+| User-controlled data in arithmetic expression (`cpp/tainted-arithmetic`) | More results | The query is extended to analyze increment, decrement, addition assignment and subtraction assignment operations. |
+| Variable used in its own initializer (`cpp/use-in-own-initializer`) | Fewer false positive results | Results where a macro is used to indicate deliberate uninitialization are now excluded |
+|Uncontrolled data in arithmetic expression (`cpp/uncontrolled-arithmetic`) | More results | The query is extended to analyze increment, decrement, addition assignment and subtraction assignment operations. |
```
Comment (jbj) [inline change-notes/1.18/analysis-cpp.md]: Missing comma before "and". Also, I think "addition assignment" and "subtraction assignment" would read better if hyphenated. Without hyphens, I find this sentence hard to parse.

Same issue above, with `cpp/tainted-arithmetic`.

File: change-notes/1.18/analysis-cpp.md [hunk 3]
```diff
@@ -1,40 +1,28 @@
...
+| Nested loops with same variable (`cpp/nested-loops-with-same-variable`) | Fewer false positive results | Results where the loop variable is a member of a class or struct now account for the object. |
+| Self comparison (`cpp/comparison-of-identical-expressions`) | Fewer false positive results | Range checks of the form `x == (T)x` are no longer flagged unless they are guaranteed to have the same result on all platforms. |
+| Too few arguments to formatting function (`cpp/wrong-number-format-arguments`) | More precise results | This was previously known as "Wrong number of arguments to formatting function". It now focuses only functions calls that are missing arguments, which tend to be more severe. See the next row for the new query that reports lower-severity alerts for calls with too many arguments. In addition, both queries now understand positional format arguments as supported by some libraries, and some false positive results for custom printf-like functions have been fixed.|
```
Comment (jbj) [inline change-notes/1.18/analysis-cpp.md]: "focuses only functions" -> "focuses only _on_ functions"

File: change-notes/1.18/analysis-cpp.md [hunk 4]
```diff
@@ -1,40 +1,28 @@
...
+| Too many arguments to formatting function (`cpp/too-many-format-arguments`) | More precise results | This new query was created by splitting the old "Wrong number of arguments to formatting function" query (see row above). It reports function calls with too many arguments.  |
+| User-controlled data in arithmetic expression (`cpp/tainted-arithmetic`) | More results | The query is extended to analyze increment, decrement, addition assignment and subtraction assignment operations. |
+| Variable used in its own initializer (`cpp/use-in-own-initializer`) | Fewer false positive results | Results where a macro is used to indicate deliberate uninitialization are now excluded |
```
Comment (jbj) [inline change-notes/1.18/analysis-cpp.md]: Missing full stop.

**Comment by felicity-semmle** [comment]
Thanks for the review Jonas. Text updated accordingly 😄
